internal/db/role_test.go

```diff
@@ -0,0 +1,195 @@
+	c.Assert(err, qt.IsNil)
+
+	roleEntry, err := s.Database.AddRole(ctx, "test-role")
+	c.Assert(err, qt.IsNil)
+	c.Assert(roleEntry.UUID, qt.Not(qt.Equals), "")
+
+	_, err = s.Database.AddRole(ctx, "test-role")
+	c.Assert(errors.ErrorCode(err), qt.Equals, errors.CodeAlreadyExists)
+
+	re := dbmodel.RoleEntry{
+		Name: "test-role",
+	}
+	tx := s.Database.DB.First(&re)
+	c.Assert(tx.Error, qt.IsNil)
+	c.Assert(re.ID, qt.Equals, uint(1))
+	c.Assert(re.Name, qt.Equals, "test-role")
+	c.Assert(re.UUID, qt.Equals, uuid)
+}
+
+func (s *dbSuite) TestDatabase_GetRole(c *qt.C) {
+	uuid1 := uuid.NewString()
+	c.Patch(db.NewUUID, func() string {
+		return uuid1
+	})
+
+	err := s.Database.GetRole(context.Background(), &dbmodel.RoleEntry{
+		Name: "test-role",
+	})
+	c.Check(errors.ErrorCode(err), qt.Equals, errors.CodeUpgradeInProgress)
+
+	err = s.Database.Migrate(context.Background(), false)
+	c.Assert(err, qt.IsNil)
+
+	role := &dbmodel.RoleEntry{
+		Name: "test-role",
+	}
+	err = s.Database.GetRole(context.Background(), role)
+	c.Check(errors.ErrorCode(err), qt.Equals, errors.CodeNotFound)
+
+	roleEntry, err := s.Database.AddRole(context.TODO(), "test-role")
+	c.Assert(err, qt.IsNil)
+	c.Assert(roleEntry.UUID, qt.Equals, uuid1)
+
+	err = s.Database.GetRole(context.Background(), role)
+	c.Check(err, qt.IsNil)
+	c.Assert(role.ID, qt.Equals, uint(1))
+	c.Assert(role.Name, qt.Equals, "test-role")
+	c.Assert(role.UUID, qt.Equals, uuid1)
+
+	uuid2 := uuid.NewString()
+	c.Patch(db.NewUUID, func() string {
+		return uuid2
+	})
+
+	roleEntry, err = s.Database.AddRole(context.Background(), "test-role1")
+	c.Assert(err, qt.IsNil)
+	c.Assert(roleEntry.UUID, qt.Equals, uuid2)
+
+	role = &dbmodel.RoleEntry{
+		Name: "test-role1",
+	}
+
+	err = s.Database.GetRole(context.Background(), role)
+	c.Check(err, qt.IsNil)
+	c.Assert(role.ID, qt.Equals, uint(2))
+	c.Assert(role.Name, qt.Equals, "test-role1")
+	c.Assert(role.UUID, qt.Equals, uuid2)
+}
+
+func (s *dbSuite) TestDatabase_UpdateRole(c *qt.C) {
+	err := s.Database.UpdateRole(context.Background(), &dbmodel.RoleEntry{Name: "test-role"})
+	c.Check(errors.ErrorCode(err), qt.Equals, errors.CodeNotFound)
+
+	err = s.Database.Migrate(context.Background(), false)
+	c.Assert(err, qt.IsNil)
+
+	ge := &dbmodel.RoleEntry{
+		Name: "test-role",
+	}
+
+	err = s.Database.UpdateRole(context.Background(), ge)
+	c.Check(errors.ErrorCode(err), qt.Equals, errors.CodeNotFound)
+
+	_, err = s.Database.AddRole(context.Background(), "test-role")
+	c.Assert(err, qt.IsNil)
+
+	ge1 := &dbmodel.RoleEntry{
+		Name: "test-role",
+	}
+	err = s.Database.GetRole(context.Background(), ge1)
+	c.Assert(err, qt.IsNil)
+
+	ge1.Name = "renamed-role"
+	err = s.Database.UpdateRole(context.Background(), ge1)
+	c.Check(err, qt.IsNil)
+
+	ge2 := &dbmodel.RoleEntry{
+		Name: "renamed-role",
+	}
+	err = s.Database.GetRole(context.Background(), ge2)
+	c.Check(err, qt.IsNil)
+	c.Assert(ge2, qt.DeepEquals, ge1)
```

Comment:
how they are deep equals? At least their `UpdatedAt` should be different, right?

Reply:
Fair point! I'm not actually sure...

Reply:
worth investigating.

Reply:
I think gorm is not aware of UpdateAt. 

When we are not using gorm in the models maybe we've also lost this autoupdating mechanism for `CreatedAt, UpdatedAt`. Please investigate this or create a card

Reply:
Looking now

Reply:
Created bug.